[Bug][worker] Insert into clickhouse table successfully but process status turn to failed and log exception

**For better global communication, Please describe it in English. If you feel the description in English is not clear, then you can append the description in Chinese(just for Mandarin(CN)), thx! **
**Describe the bug**
DATASORUCE: clickhouse
My process only has one SQL like:
`INSERT INTO test
SELECT dt, count(1) as c
  FROM dau
 GROUP BY dt
 ORDER BY dt`
The process failed but the data has been successfully inserted into the table.

**To Reproduce**
Steps to reproduce the behavior, for example:
1.  Create a new process with only one SQL
`INSERT INTO test
SELECT dt, count(1) as c
  FROM dau
 GROUP BY dt
 ORDER BY dt`
2. Save and run it.
3. See error below
[INFO] 2021-04-19 17:49:57.364  - [taskAppId=TASK-2-9-10]:[115] - create dir success /tmp/dolphinscheduler/exec/process/1/2/9/10
[INFO] 2021-04-19 17:49:57.367  - [taskAppId=TASK-2-9-10]:[88] - sql task params {"postStatements":[],"connParams":"","receiversCc":"","udfs":"","type":"CLICKHOUSE","title":"test","sql":"INSERT INTO test\nSELECT dt, count(1) as c\n  FROM dau\n GROUP BY dt\n ORDER BY dt\n","preStatements":[],"sqlType":"0","receivers":"xxx@xxx.com","datasource":1,"showType":"TABLE","localParams":[]}
[INFO] 2021-04-19 17:49:57.369  - [taskAppId=TASK-2-9-10]:[104] - Full sql parameters: SqlParameters{type='CLICKHOUSE', datasource=1, sql='INSERT INTO test
SELECT dt, count(1) as c
  FROM dau
 GROUP BY dt
 ORDER BY dt
', sqlType=0, udfs='', showType='TABLE', connParams='', title='test', receivers='xxx@xxx.com', receiversCc='', preStatements=[], postStatements=[]}
[INFO] 2021-04-19 17:49:57.369  - [taskAppId=TASK-2-9-10]:[105] - sql type : CLICKHOUSE, datasource : 1, sql : INSERT INTO test
SELECT dt, count(1) as c
  FROM dau
 GROUP BY dt
 ORDER BY dt
 , localParams : [],udfs : ,showType : TABLE,connParams : 
[INFO] 2021-04-19 17:49:57.369  - [taskAppId=TASK-2-9-10]:[176] - SQL title : test
[INFO] 2021-04-19 17:49:57.370  - [taskAppId=TASK-2-9-10]:[498] - after replace sql , preparing : INSERT INTO test
SELECT dt, count(1) as c
  FROM dau
 GROUP BY dt
 ORDER BY dt

[INFO] 2021-04-19 17:49:57.370  - [taskAppId=TASK-2-9-10]:[503] - Sql Params are replaced sql , parameters:
[INFO] 2021-04-19 17:49:57.370  - [taskAppId=TASK-2-9-10]:[52] - can't find udf function resource
[INFO] 2021-04-19 17:49:57.381  - [taskAppId=TASK-2-9-10]:[414] - prepare statement replace sql : ru.yandex.clickhouse.ClickHousePreparedStatementImpl@65fc3a50 
[ERROR] 2021-04-19 17:49:57.388  - [taskAppId=TASK-2-9-10]:[243] - execute sql error
java.lang.NullPointerException: null
	at org.apache.dolphinscheduler.server.worker.task.sql.SqlTask.resultProcess(SqlTask.java:258)
	at org.apache.dolphinscheduler.server.worker.task.sql.SqlTask.executeFuncAndSql(SqlTask.java:233)
	at org.apache.dolphinscheduler.server.worker.task.sql.SqlTask.handle(SqlTask.java:139)
	at org.apache.dolphinscheduler.server.worker.runner.TaskExecuteThread.run(TaskExecuteThread.java:133)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[ERROR] 2021-04-19 17:49:57.388  - [taskAppId=TASK-2-9-10]:[145] - sql task error
java.lang.RuntimeException: execute sql error
	at org.apache.dolphinscheduler.server.worker.task.sql.SqlTask.executeFuncAndSql(SqlTask.java:244)
	at org.apache.dolphinscheduler.server.worker.task.sql.SqlTask.handle(SqlTask.java:139)
	at org.apache.dolphinscheduler.server.worker.runner.TaskExecuteThread.run(TaskExecuteThread.java:133)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)


**Expected behavior**
A clear and concise description of what you expected to happen.

Data inserted into the table and process status should be successful.

**Screenshots**
If applicable, add screenshots to help explain your problem.
![1618827014390](https://user-images.githubusercontent.com/4523603/115220161-d04fd500-a13a-11eb-9a5e-471897e4180a.jpg)
![1618827054023](https://user-images.githubusercontent.com/4523603/115220193-d776e300-a13a-11eb-95ce-61b719789c2e.jpg)
![1618827080868](https://user-images.githubusercontent.com/4523603/115220240-e2ca0e80-a13a-11eb-9b79-547627751944.jpg)


**Which version of Dolphin Scheduler:**
 -1.3.5

**Additional context**
Add any other context about the problem here.
No matter deployed by docker or k8s, both met this problem.

**Requirement or improvement**
- Please describe about your requirements or improvement suggestions.

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/sql/SqlTask.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/sql/SqlTask.java
@@ -328,11 +328,12 @@ public class SqlTask extends AbstractTask {
             }
         }
 
-        if (sqlParameters.getSendEmail() == null || sqlParameters.getSendEmail()) {
-            sendAttachment(sqlParameters.getGroupId(), StringUtils.isNotEmpty(sqlParameters.getTitle()) ? sqlParameters.getTitle() : taskExecutionContext.getTaskName() + " query result sets",
-                    JSONUtils.toJsonString(resultJSONArray));
-        }
         String result = JSONUtils.toJsonString(resultJSONArray);
+        if (sqlParameters.getSendEmail() == null || sqlParameters.getSendEmail()) {
+            sendAttachment(sqlParameters.getGroupId(), StringUtils.isNotEmpty(sqlParameters.getTitle())
+                            ? sqlParameters.getTitle()
+                            : taskExecutionContext.getTaskName() + " query result sets", result);
+        }
         logger.debug("execute sql result : {}", result);
         return result;
     }

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/sql/SqlTask.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/sql/SqlTask.java
@@ -304,34 +304,36 @@ public class SqlTask extends AbstractTask {
      */
     private String resultProcess(ResultSet resultSet) throws Exception {
         ArrayNode resultJSONArray = JSONUtils.createArrayNode();
-        ResultSetMetaData md = resultSet.getMetaData();
-        int num = md.getColumnCount();
+        if (resultSet != null) {
+            ResultSetMetaData md = resultSet.getMetaData();
+            int num = md.getColumnCount();
 
-        int rowCount = 0;
+            int rowCount = 0;
 
-        while (rowCount < LIMIT && resultSet.next()) {
-            ObjectNode mapOfColValues = JSONUtils.createObjectNode();
-            for (int i = 1; i <= num; i++) {
-                mapOfColValues.set(md.getColumnLabel(i), JSONUtils.toJsonNode(resultSet.getObject(i)));
+            while (rowCount < LIMIT && resultSet.next()) {
+                ObjectNode mapOfColValues = JSONUtils.createObjectNode();
+                for (int i = 1; i <= num; i++) {
+                    mapOfColValues.set(md.getColumnLabel(i), JSONUtils.toJsonNode(resultSet.getObject(i)));
+                }
+                resultJSONArray.add(mapOfColValues);
+                rowCount++;
             }
-            resultJSONArray.add(mapOfColValues);
-            rowCount++;
-        }
-        String result = JSONUtils.toJsonString(resultJSONArray);
-        logger.debug("execute sql result : {}", result);
 
-        int displayRows = sqlParameters.getDisplayRows() > 0 ? sqlParameters.getDisplayRows() : Constants.DEFAULT_DISPLAY_ROWS;
-        displayRows = Math.min(displayRows, resultJSONArray.size());
-        logger.info("display sql result {} rows as follows:", displayRows);
-        for (int i = 0; i < displayRows; i++) {
-            String row = JSONUtils.toJsonString(resultJSONArray.get(i));
-            logger.info("row {} : {}", i + 1, row);
+            int displayRows = sqlParameters.getDisplayRows() > 0 ? sqlParameters.getDisplayRows() : Constants.DEFAULT_DISPLAY_ROWS;
+            displayRows = Math.min(displayRows, resultJSONArray.size());
+            logger.info("display sql result {} rows as follows:", displayRows);
+            for (int i = 0; i < displayRows; i++) {
+                String row = JSONUtils.toJsonString(resultJSONArray.get(i));
+                logger.info("row {} : {}", i + 1, row);
+            }
         }
 
         if (sqlParameters.getSendEmail() == null || sqlParameters.getSendEmail()) {
             sendAttachment(sqlParameters.getGroupId(), StringUtils.isNotEmpty(sqlParameters.getTitle()) ? sqlParameters.getTitle() : taskExecutionContext.getTaskName() + " query result sets",
                     JSONUtils.toJsonString(resultJSONArray));
         }
+        String result = JSONUtils.toJsonString(resultJSONArray);
+        logger.debug("execute sql result : {}", result);
         return result;
     }
 

--- a/dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/sql/SqlTaskTest.java
+++ b/dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/sql/SqlTaskTest.java
@@ -20,6 +20,7 @@ package org.apache.dolphinscheduler.server.worker.task.sql;
 import org.apache.dolphinscheduler.common.Constants;
 import org.apache.dolphinscheduler.common.utils.ParameterUtils;
 import org.apache.dolphinscheduler.dao.AlertDao;
+import org.apache.dolphinscheduler.dao.datasource.BaseDataSource;
 import org.apache.dolphinscheduler.remote.command.alert.AlertSendResponseCommand;
 import org.apache.dolphinscheduler.server.entity.SQLTaskExecutionContext;
 import org.apache.dolphinscheduler.server.entity.TaskExecutionContext;
@@ -30,6 +31,8 @@ import org.apache.dolphinscheduler.service.bean.SpringApplicationContext;
 import java.sql.Connection;
 import java.sql.DriverManager;
 import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
 import java.util.Date;
 
 import org.junit.Assert;
@@ -48,7 +51,8 @@ import org.slf4j.LoggerFactory;
  *  sql task test
  */
 @RunWith(PowerMockRunner.class)
-@PrepareForTest(value = {SqlTask.class, DriverManager.class, SpringApplicationContext.class, ParameterUtils.class})
+@PrepareForTest(value = {SqlTask.class, DriverManager.class, SpringApplicationContext.class,
+        ParameterUtils.class, AlertSendResponseCommand.class, BaseDataSource.class})
 public class SqlTaskTest {
 
     private static final Logger logger = LoggerFactory.getLogger(SqlTaskTest.class);
@@ -124,6 +128,25 @@ public class SqlTaskTest {
         PowerMockito.when(mockResponseCommand.getResStatus()).thenReturn(true);
         PowerMockito.when(alertClientService.sendAlert(0, "null query result sets", "[]")).thenReturn(mockResponseCommand);
         String result = Whitebox.invokeMethod(sqlTask, "resultProcess", null);
+        Assert.assertNotNull(result);
+    }
+
+    @Test
+    public void testResultProcess02() throws Exception {
+        // test input not null
+        ResultSet resultSet = PowerMockito.mock(ResultSet.class);
+        ResultSetMetaData mockResultMetaData = PowerMockito.mock(ResultSetMetaData.class);
+        PowerMockito.when(resultSet.getMetaData()).thenReturn(mockResultMetaData);
+        PowerMockito.when(mockResultMetaData.getColumnCount()).thenReturn(2);
+        PowerMockito.when(resultSet.next()).thenReturn(true);
+        PowerMockito.when(resultSet.getObject(Mockito.anyInt())).thenReturn(1);
+        PowerMockito.when(mockResultMetaData.getColumnLabel(Mockito.anyInt())).thenReturn("a");
+
+        AlertSendResponseCommand mockResponseCommand = PowerMockito.mock(AlertSendResponseCommand.class);
+        PowerMockito.when(mockResponseCommand.getResStatus()).thenReturn(true);
+        PowerMockito.when(alertClientService.sendAlert(Mockito.anyInt(), Mockito.anyString(), Mockito.anyString())).thenReturn(mockResponseCommand);
+
+        String result = Whitebox.invokeMethod(sqlTask, "resultProcess", resultSet);
         Assert.assertNotNull(result);
     }
 }

--- a/dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/sql/SqlTaskTest.java
+++ b/dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/sql/SqlTaskTest.java
@@ -20,6 +20,7 @@ package org.apache.dolphinscheduler.server.worker.task.sql;
 import org.apache.dolphinscheduler.common.Constants;
 import org.apache.dolphinscheduler.common.utils.ParameterUtils;
 import org.apache.dolphinscheduler.dao.AlertDao;
+import org.apache.dolphinscheduler.remote.command.alert.AlertSendResponseCommand;
 import org.apache.dolphinscheduler.server.entity.SQLTaskExecutionContext;
 import org.apache.dolphinscheduler.server.entity.TaskExecutionContext;
 import org.apache.dolphinscheduler.server.worker.task.TaskProps;
@@ -39,6 +40,7 @@ import org.mockito.Mockito;
 import org.powermock.api.mockito.PowerMockito;
 import org.powermock.core.classloader.annotations.PrepareForTest;
 import org.powermock.modules.junit4.PowerMockRunner;
+import org.powermock.reflect.Whitebox;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -112,6 +114,16 @@ public class SqlTaskTest {
         PowerMockito.when(ParameterUtils.replaceScheduleTime(Mockito.any(), Mockito.any())).thenReturn("insert into tb_1 values('1','2')");
 
         sqlTask.handle();
-        Assert.assertEquals(Constants.EXIT_CODE_SUCCESS,sqlTask.getExitStatusCode());
+        Assert.assertEquals(Constants.EXIT_CODE_SUCCESS, sqlTask.getExitStatusCode());
+    }
+
+    @Test
+    public void testResultProcess() throws Exception {
+        // test input null and will not throw a exception
+        AlertSendResponseCommand mockResponseCommand = PowerMockito.mock(AlertSendResponseCommand.class);
+        PowerMockito.when(mockResponseCommand.getResStatus()).thenReturn(true);
+        PowerMockito.when(alertClientService.sendAlert(0, "null query result sets", "[]")).thenReturn(mockResponseCommand);
+        String result = Whitebox.invokeMethod(sqlTask, "resultProcess", null);
+        Assert.assertNotNull(result);
     }
 }

--- a/pom.xml
+++ b/pom.xml
@@ -963,6 +963,7 @@
                         <include>**/server/worker/task/TaskManagerTest.java</include>
                         <include>**/server/worker/task/AbstractCommandExecutorTest.java</include>
                         <include>**/server/worker/task/ShellTaskReturnTest.java</include>
+                        <include>**/server/worker/task/sql/SqlTaskTest.java</include>
                         <include>**/server/worker/EnvFileTest.java</include>
                         <include>**/server/worker/runner/TaskExecuteThreadTest.java</include>
                         <include>**/server/worker/runner/WorkerManagerThreadTest.java</include>

